src/metabase/models/timeline.clj

```diff
@@ -35,6 +37,65 @@
+  well as `all?`. By default, will return only unarchived events, unless `all?` is truthy and will return all events
+  regardless of archive state."
+  [timeline-ids {:events/keys [all? start end]}]
+  (let [timeline-icon {:select [[(hsql/call :coalesce :icon DefaultIcon) :tl_icon]
+                                [:id :tl_id]]
+                       :from   [Timeline]}
+        query         {:select    [[(hsql/call :coalesce :icon :tl_icon) :icon]
+                                   :description :archived :timezone :time_matters :name
+                                   :timeline_id :creator_id :updated_at :id :timestamp :created_at]
+                       :from      [TimelineEvent]
+                       :where     [:and
+                               ;; in our collections
+                               [:in :timeline_id timeline-ids]
+                               (when-not all?
+                                 [:= :archived false])
+                               (when (or start end)
+                                 [:or
+                                  ;; absolute time in bounds
+                                  [:and
+                                   [:= :time_matters true]
+                                   ;; less than or equal?
+                                   (when start
+                                     [:<= start :timestamp])
+                                   (when end
+                                     [:<= :timestamp end])]
+                                  ;; non-specic time in bounds
+                                  [:and
+                                   [:= :time_matters false]
+                                   (when start
+                                     [:<= (hx/->date start) (hx/->date :timestamp)])
+                                   (when end
+                                     [:<= (hx/->date :timestamp) (hx/->date end)])]])]
+                       :left-join [[timeline-icon :tl] [:= :timeline_id :tl_id]]}]
+    (hydrate (map #(dissoc % :icon_2 :tl_id :tl_icon) (db/query query)) :creator)))
+
+(defn include-events
+  "Include events on `timelines` passed in. Options are optional and include whether to return unarchived events or all
+  events regardless of archive status (`all?`), and `start` and `end` parameters for events."
+  [timelines options]
+  (if-not (seq timelines)
+    []
+    (let [timeline-id->events (->> (fetch-events (map :id timelines) options)
+                                   (group-by :timeline_id))]
+      (for [{:keys [id] :as timeline} timelines]
+        (let [events (timeline-id->events id)]
+          (when timeline
+            (assoc timeline :events (if events events []))))))))
+
+(defn include-events-singular
+  "Similar to [[include-events]] but allows for passing a single timeline not in a collection."
+  ([timeline] (include-events-singular timeline {}))
+  ([timeline options]
+   (first (include-events [timeline] options))))
+
```

Comment:
I moved these few functions into timeline to avoid a circular dependency issue with timeline-event (timeline-event requires timeline for DefaultIcon)